### PR TITLE
[Inductor][CPU] Guard compiled integer div by zero

### LINKED ISSUE #143649: Floating Point Exception (core dumped) when running floordiv/remainder/fmod under torch.compile
### 🐛 Describe the bug

It is likely a division-by-zero problem. 
Under eager mode, these APIs do not throw the `ZeroDivisionError` exception instead of a floating point error.

Here is the code to reproduce:
```

import torch


@torch.compile
def div(input,value):
    return torch.Tensor.floor_divide_(input,value)  # change the API to torch.fmod or torch.remainder will also lead to the same error

input = torch.tensor([2,5])
value = torch.tensor([0])
div(input, value)

```

### Versions

Collecting environment information...
PyTorch version: 2.6.0a0+gitdeb1da1
Is debug build: False
CUDA used to build PyTorch: 12.4
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.4 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.31.2
Libc version: glibc-2.35

Python version: 3.11.10 | packaged by conda-forge | (main, Oct 16 2024, 01:27:36) [GCC 13.3.0] (64-bit runtime)
Python platform: Linux-5.14.0-427.37.1.el9_4.x86_64-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.4.131
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090

Nvidia driver version: 560.35.03
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      43 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             64
On-line CPU(s) list:                0-63
Vendor ID:                          AuthenticAMD
Model name:                         AMD Ryzen Threadripper PRO 3975WX 32-Cores
CPU family:                         23
Model:                              49
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          1
Stepping:                           0
Frequency boost:                    enabled
CPU max MHz:                        4368.1641
CPU min MHz:                        2200.0000
BogoMIPS:                           7000.73
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 hw_pstate ssbd mba ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold avic v_vmsave_vmload vgif v_spec_ctrl umip rdpid overflow_recov succor smca sev sev_es
Virtualization:                     AMD-V
L1d cache:                          1 MiB (32 instances)
L1i cache:                          1 MiB (32 instances)
L2 cache:                           16 MiB (32 instances)
L3 cache:                           128 MiB (8 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-63
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Mitigation; untrained return thunk; SMT enabled with STIBP protection
Vulnerability Spec rstack overflow: Mitigation; Safe RET
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Retpolines, IBPB conditional, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] numpy==2.1.2
[pip3] nvidia-cublas-cu12==12.4.5.8
[pip3] nvidia-cuda-cupti-cu12==12.4.127
[pip3] nvidia-cuda-nvrtc-cu12==12.4.127
[pip3] nvidia-cuda-runtime-cu12==12.4.127
[pip3] nvidia-cudnn-cu12==9.1.0.70
[pip3] nvidia-cufft-cu12==11.2.1.3
[pip3] nvidia-curand-cu12==10.3.5.147
[pip3] nvidia-cusolver-cu12==11.6.1.9
[pip3] nvidia-cusparse-cu12==12.3.1.170
[pip3] nvidia-nccl-cu12==2.21.5
[pip3] nvidia-nvjitlink-cu12==12.4.127
[pip3] nvidia-nvtx-cu12==12.4.127
[pip3] optree==0.13.0
[pip3] torch==2.6.0a0+gitdeb1da1
[pip3] torchaudio==2.5.1+cu124
[pip3] torchelastic==0.2.2
[pip3] torchvision==0.20.1+cu124
[pip3] triton==3.1.0
[conda] magma-cuda124             2.6.1                         1    pytorch
[conda] mkl                       2025.0.0           h901ac74_941    conda-forge
[conda] mkl-include               2025.0.0           hf2ce2f3_941    conda-forge
[conda] numpy                     2.1.2           py311h71ddf71_0    conda-forge
[conda] nvidia-cublas-cu12        12.4.5.8                 pypi_0    pypi
[conda] nvidia-cuda-cupti-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-nvrtc-cu12    12.4.127                 pypi_0    pypi
[conda] nvidia-cuda-runtime-cu12  12.4.127                 pypi_0    pypi
[conda] nvidia-cudnn-cu12         9.1.0.70                 pypi_0    pypi
[conda] nvidia-cufft-cu12         11.2.1.3                 pypi_0    pypi
[conda] nvidia-curand-cu12        10.3.5.147               pypi_0    pypi
[conda] nvidia-cusolver-cu12      11.6.1.9                 pypi_0    pypi
[conda] nvidia-cusparse-cu12      12.3.1.170               pypi_0    pypi
[conda] nvidia-nccl-cu12          2.21.5                   pypi_0    pypi
[conda] nvidia-nvjitlink-cu12     12.4.127                 pypi_0    pypi
[conda] nvidia-nvtx-cu12          12.4.127                 pypi_0    pypi
[conda] optree                    0.13.0                   pypi_0    pypi
[conda] torch                     2.6.0a0+gitdeb1da1          pypi_0    pypi
[conda] torchaudio                2.5.1+cu124              pypi_0    pypi
[conda] torchelastic              0.2.2                    pypi_0    pypi
[conda] torchvision               0.20.1+cu124             pypi_0    pypi
[conda] triton                    3.1.0                    pypi_0    pypi

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #186021

CPU Inductor lowered integer floor division and trunc division to raw C++
`/` and `%` expressions. When a compiled kernel saw a zero divisor, the
hardware raised SIGFPE before PyTorch could turn the condition into a Python
exception. `torch.fmod` and `torch.remainder` already used helper routines that
record integer divide errors through the per-kernel error flag and throw after
kernel execution.

Route scalar and vector integer floor/trunc division through matching helper
routines in the C++ prefix. The helpers check for zero divisors, set the
existing Inductor integer-div-error flag, and return a placeholder value so the
kernel can complete and report `RuntimeError("ZeroDivisionError")` through the
same path used by remainder/fmod. For nonzero
divisors, the helpers preserve existing raw division behavior.

This keeps the fix localized to CPU C++ codegen instead of adding Python-side
prechecks, which would miss generated-code paths and add extra tensor work.

Fixes #143649
Generated by my agent

Benchmark Results:
A small CPU int64 non-zero divisor microbenchmark was run with
`torch.set_num_threads(1)`, tensor shape `(1024, 1024)`, 30 timed compiled
iterations after warmup.

Before:
- floor_divide median=0.002108s mean=0.002559s min=0.002075s
- divide_trunc median=0.002092s mean=0.002570s min=0.002072s

After:
- floor_divide median=0.002088s mean=0.002656s min=0.002067s
- divide_trunc median=0.002083s mean=0.002152s min=0.002062s

Test Plan:
- Standalone Python script compiling and running floor_divide, divide_trunc,
  fmod, and remainder on CPU int32/int64 tensors with a zero divisor lane: all
  raised RuntimeError containing ZeroDivisionError.
- Standalone Python script comparing compiled vs eager non-zero CPU int32/int64
  floor_divide and divide_trunc over positive and negative divisors: passed.
- `lintrunner --take CLANGFORMAT,PYFMT,RUFF,FLAKE8,TYPEIGNORE -a`: passed.
- `git diff --check`: passed.
- `lintrunner -a`: failed in CLANGTIDY due to local/pre-existing header issues
  including missing `omp.h` and unrelated warnings in `cpp_prefix.h`.
- `pytest -q test/inductor/test_torchinductor.py -k 'floor_divide_zero_divisor_cpu_inductor_raises_error or trunc_divide_zero_divisor_cpu_inductor_raises_error or fmod_uint8_zero_divisor_cpu_inductor_raises_error or remainder_uint8_zero_divisor_cpu_inductor_raises_error'`:
  failed during collection because installed torchvision is incompatible with
  this checkout (`operator torchvision::nms does not exist`).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says I rebased onto latest main and resubmitted. I also narrowed the trunc-div helper to only handle zero divisors so nonzero divisor behavior stays unchanged. The prior new CI failure was in HuggingFace DebertaV2ForMaskedLM eager import/TorchScript setup before Inductor ran; fresh CI is now running on the updated branch.

**Comment by jansel** [comment]
My agent says I corrected the PR description to match the current implementation: the helpers preserve existing raw division behavior for nonzero divisors; the previously mentioned signed min/-1 special case was removed.

**Comment by jansel** [comment]
My agent says I rebased this PR onto current main and refreshed it to focus on truncating integer division, since the floor-division fixes landed separately. During review I reproduced a concurrency race in the prior process-global error flag; this update now uses a per-invocation atomic passed explicitly through scalar/vector helpers and every C++ template path, including an outer-owned flag for BMM, with checks after OpenMP regions.

Validation passed for int32/int64 floor, trunc, fmod, and remainder zero-divisor handling, vector tails, float fmod, OpenMP deferral, eight simultaneous C++ wrapper calls, CPU flex attention, the C++ micro-GEMM BMM template, focused linters, byte-compilation, and diff checks. Full lint is clean apart from local CLANGTIDY being unable to find omp.h. Multiple fresh max-reasoning reviews found no blocking issues.

**Comment by jansel** [comment]
@claude review these changes

**Comment by jansel** [comment]
My agent says this PR is being closed as part of a cleanup of open agentic PRs to focus review bandwidth on a smaller set; the linked issue remains open.
